### PR TITLE
HPCC-21259 generateGlobalId workunit option to generate GlobalIds

### PR DESCRIPTION
**Pull request by afishbeck**
If option generateGlobalId is set and a GlobalId is not sent via
HTTP header, then roxie will generate a new unique GlobalId.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by richardkchapman** [comment]
@AttilaVamos Any ideas what failed on smoketest?

**Comment by AttilaVamos** [comment]
@afishbeck The problem is:
```
CMake Error at cmake_modules/commonSetup.cmake:480 (add_library):
  Cannot find source file:

    /mnt/disk1/home/vamosax/smoketest/PR-12311/HPCC-Platform/system/globalid/lnuid.cpp
```
And if I see well, the lnuid.h also.
Somehow, the most important part is missing from the report. Please, give me some time to check and fix Smoketest.

**Comment by richardkchapman** [comment]
I suspect Tony forgot to add them

@afishbeck

**Comment by afishbeck** [comment]
Added missing files.  Waiting for smoke test to confirm.

**Comment by richardkchapman** [comment]
@wangkx Please review

File: system/globalid/lnuid.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+    void getUniqueIdRange(time_t start, time_t end, ln_uid_t &uid_start, ln_uid_t &uid_end);
+    void getUniqueIdDateRange(const char *start, const char *end, ln_uid_t &uid_start, ln_uid_t &uid_end);
+
+
+    bool sameUniqueId(const ln_uid_t &uid1, const ln_uid_t &uid2);
+    void copyUniqueId(ln_uid_t &to, const ln_uid_t &from);
+
+	int get_utc_offset();
```
Comment (wangkx) [inline system/globalid/lnuid.h]: wrong indent?

File: system/globalid/lnuid.cpp [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+    ln_uid_t &createUniqueId(ln_uid_t &out){
+
+        unsigned long long  int millisecondsSinceEpochTrim;
+#ifdef __linux__
+        struct timeval tv;
+        gettimeofday(&tv, NULL);
+
+        millisecondsSinceEpochTrim =
+        (unsigned long long int)(tv.tv_sec) * 1000 +
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: indent?

File: system/globalid/lnuid.cpp [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+        HCRYPTPROV hProvider;
+        CryptAcquireContext(&hProvider, 0, 0, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT);
+        CryptGenRandom(hProvider, random_byte_count, randomdata);
+#else
+
+#endif
+
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            out[i] = timedata[j];
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: Why do you need both i and j (they are the same)?

File: system/globalid/lnuid.cpp [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            }
+
+            assert(carry == 0);
+            length = i;
+            counter++;
+        }
+        // Skip leading zeroes in base58 result.
+        vector<unsigned char>::iterator it = b58.begin() + (size - length);
+        while (it != b58.end() && *it == 0)
+        it++;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: indent?

File: system/globalid/lnuid.cpp [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        // Translate the result into a string.
+        string str;
+        str.reserve(zeroes + (b58.end() - it));
+        str.assign(zeroes, '1');
+        while (it != b58.end())
+        str += pszBase58[*(it++)];
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: indent?

File: system/globalid/lnuid.cpp [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+        return str;
+
+    }
+
+    ln_uid_t &uniqueIdFromString(const char* uid, ln_uid_t &out){
+
+        // Skip leading spaces. psz
+        while (*uid && isspace(*uid))
+        uid++;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+            uid++;
+        }
+        // Allocate enough space in big-endian base256 representation.
+        vector<unsigned char> b256(strlen(uid) * 733 / 1000 + 1); // log(58) / log(256), rounded up.
+        // Process the characters.
+        while (*uid && !isspace(*uid)) {
+            // Decode base58 character
+            const char* ch = strchr(pszBase58, *uid);
+            if (ch == NULL)
+            return out;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+                carry += 58 * (*it);
+                *it = carry % 256;
+                carry /= 256;
+            }
+            assert(carry == 0);
+            uid++;
+        }
+        // Skip trailing spaces.
+        while (isspace(*uid))
+        uid++;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        if (*uid != 0)
+        return out;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+        // Skip leading zeroes in b256.
+        vector<unsigned char>::iterator it = b256.begin();
+        while (it != b256.end() && *it == 0)
+        it++;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+        vector<unsigned char> vch;
+        // Copy result into output vector.
+        vch.reserve(zeroes + (b256.end() - it));
+        vch.assign(zeroes, 0x00);
+        while (it != b256.end())
+        vch.push_back(*(it++));
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+        time_t seconds = (time_t)(timepart/1000);
+
+        return seconds;
+
+    }
+
+    void getUniqueIdRange(time_t start, time_t end, ln_uid_t &uid_start, ln_uid_t &uid_end){
+
+        unsigned long long  int startMilliSecondsSinceEpochTrim =
+        (unsigned long long int)(start) * 1000;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+        unsigned long long  int endMilliSecondsSinceEpochTrim =
+        (unsigned long long int)(end) * 1000;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: Intent?

File: system/globalid/lnuid.cpp [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+
+        unsigned char blankedRandomdata[random_byte_count], filledRandomData[random_byte_count];
+
+        for(unsigned int i=0; i <random_byte_count; i++) {
+            blankedRandomdata[i] = 0x00;
+            filledRandomData[i] = 0xFF;
+        }
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            uid_start[i] = startTimeData[j] ;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: uid_start[i] = startTimeData[i];

File: system/globalid/lnuid.cpp [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+        }
+        for(unsigned int i=time_byte_count, j=0; i <uid_size; i++, j++) {
+            uid_start[i] = blankedRandomdata[j] ;
+        }
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            uid_end[i] = endTimeData[j] ;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: uid_end[i] = endTimeData[i] ;

File: system/globalid/lnuid.cpp [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+        end_time = mktime(&endDateTime) ;
+        cout << "Start time:" << "UTC OFFSET:"<< get_utc_offset () << ":" <<start << ":" << start_time << endl;
+        cout << "End time:" << end << ":" << end_time << endl;
+
+        struct tm tm;
+        char sb[100];
+#ifdef __linux__
+        localtime_r(&start_time, &tm);
+        sprintf(sb, "now: %4d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+        tm.tm_hour, tm.tm_min, tm.tm_sec);
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+#elif _WIN32
+        localtime_s(&tm, &start_time);
+        sprintf_s(sb, "now: %4d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+        tm.tm_hour, tm.tm_min, tm.tm_sec);
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

File: system/globalid/lnuid.cpp [hunk 18]
```diff
@@ -0,0 +1,403 @@
...
+#else
+
+#endif
+
+        gmtime_hours = timeptr.tm_hour;
+
+        /* if the local time is the "day before" the UTC, subtract 24 hours
+from the hours to get the UTC offset */
+        if (timeptr.tm_mday < 2)
+        gmtime_hours -= 24;
```
Comment (wangkx) [inline system/globalid/lnuid.cpp]: intent?

**Review by wangkx — COMMENTED** [review]
@afishbeck several minor comments

File: system/globalid/lnuid.h (hunk 1 above)
Comment (afishbeck) [reply system/globalid/lnuid.h]: the indent issues may have been the way I copied Amol's code.  This code is used to generate IDs in LN we are open sourcing it.  I'll take a look at the original.

File: system/globalid/lnuid.cpp (hunk 3 above)
Comment (afishbeck) [reply system/globalid/lnuid.cpp]: Agreed.  Copied from Amol, but I'll change it.

**Comment by afishbeck** [comment]
Need to fix that build issue due to rebasing onto prior change.

**Comment by afishbeck** [comment]
Fixed rebase build issue.
@wangkx please re-review.

File: system/globalid/lnuid.cpp (hunk 16 above)
Comment (wangkx) [reply system/globalid/lnuid.cpp]: Missed this one?

File: system/globalid/lnuid.cpp (hunk 17 above)
Comment (wangkx) [reply system/globalid/lnuid.cpp]: Another one?

**Review by wangkx — COMMENTED** [review]
@afishbeck 2 minors

**Comment by afishbeck** [comment]
@wangkx updated.  Please check again.

**Comment by wangkx** [comment]
@afishbeck looks good.

**Comment by afishbeck** [comment]
@richardkchapman please merge.

**Pull request merged**